v201609/campaign_extension.go

```diff
@@ -0,0 +1,155 @@
+	for token, err := dec.Token(); err == nil; token, err = dec.Token() {
+		if err != nil {
+			return err
+		}
+		switch start := token.(type) {
+		case xml.StartElement:
+			switch start.Name.Local {
+			case "platformRestrictions":
+				if err := dec.DecodeElement(&s.PlatformRestrictions, &start); err != nil {
+					return err
+				}
+			case "extensions":
+				extension, err := extensionsUnmarshalXML(dec, start)
+				if err != nil {
+					return err
+				}
+				s.Extensions = append(s.Extensions, extension)
+			}
+		}
+	}
+	return nil
+}
+
+type CampaignExtensionSettingOperations map[string][]CampaignExtensionSetting
+
+// https://developers.google.com/adwords/api/docs/reference/v201609/CampaignExtensionSettingService#query
+func (s *CampaignExtensionSettingService) Query(query string) (settings []CampaignExtensionSetting, totalCount int64, err error) {
+	respBody, err := s.Auth.request(
+		campaignExtensionSettingUrl,
+		"query",
+		AWQLQuery{
+			XMLName: xml.Name{
+				Space: baseUrl,
+				Local: "query",
+			},
+			Query: query,
+		},
+	)
+	if err != nil {
+		return
+	}
+
+	getResp := struct {
+		Size     int64                      `xml:"rval>totalNumEntries"`
+		Settings []CampaignExtensionSetting `xml:"rval>entries"`
+	}{}
+
+	err = xml.Unmarshal([]byte(respBody), &getResp)
+	if err != nil {
+		return
+	}
+	return getResp.Settings, getResp.Size, err
+}
+
+// https://developers.google.com/adwords/api/docs/reference/v201609/CampaignExtensionSettingService#mutate
+func (s *CampaignExtensionSettingService) Mutate(settingsOperations CampaignExtensionSettingOperations) (settings []CampaignExtensionSetting, err error) {
+	type settingOperations struct {
+		Action  string                   `xml:"operator"`
+		Setting CampaignExtensionSetting `xml:"operand"`
+	}
+	operations := []settingOperations{}
+	for action, settings := range settingsOperations {
+		for _, setting := range settings {
+			if err = identifyCampaignExtention(&setting); err != nil {
+				return settings, err
+			}
+			operations = append(operations,
+				settingOperations{
+					Action:  action,
+					Setting: setting,
+				},
+			)
+		}
+	}
+	mutation := struct {
+		XMLName xml.Name
+		Ops     []settingOperations `xml:"operations"`
+	}{
+		XMLName: xml.Name{
+			Space: baseUrl,
+			Local: "mutate",
+		},
+		Ops: operations,
+	}
+
+	respBody, err := s.Auth.request(campaignExtensionSettingUrl, "mutate", mutation)
+	if err != nil {
+		return settings, err
+	}
+	mutateResp := struct {
+		Settings []CampaignExtensionSetting `xml:"rval>value"`
+	}{}
+	err = xml.Unmarshal(respBody, &mutateResp)
+	if err != nil {
+		return settings, err
+	}
+
+	return mutateResp.Settings, err
+}
+
+func identifyCampaignExtention(setting *CampaignExtensionSetting) (err error) {
+	switch setting.ExtensionType {
+	case "CALL":
+		for _, ext := range setting.ExtensionSetting.Extensions {
+			item := getCallFeedItem(ext.(map[string]interface{}))
```

Comment:
we should be able to utilize reflect here and remove the need for getCallFeedItem. Probably can abstract this function to be used by both campaign and adgroup since it is the same logic.